File: lib/wip/doctor.rb
# frozen_string_literal: true

require 'open3'

module Wip
  # Runs environment diagnostics and reports pass/warn/fail results.
  class Doctor
    Result = Data.define(:level, :message)

    def initialize(loader:, resolver: CommandResolver.new, environment: Environment.new,
                   compose_resolver: CommandResolver.new(candidates: [], label: 'compose command',
                                                         install_hint: ComposeBridge::INSTALL_HINT))
      @loader = loader
      @resolver = resolver
      @environment = environment
      @compose_resolver = compose_resolver
    end

    def call
      results = []
      results << result(@environment.wsl2? ? :ok : :fail, *wsl2_messages)
      results << interop_result unless @environment.windows?
      results << Result.new(:ok, "Architecture: #{@environment.architecture}")
      check_config(load_config(results), results)
      results << result(command_available?('git') ? :ok : :warn, 'Git is available',
                        'Git is not available to the WSLC build environment')
      results
    end

    private

    def wsl2_messages
      return ['WSL2 is available', 'WSL2 is not available'] if @environment.windows?

      ['Running on WSL2', 'Not running on WSL2']
    end

    def interop_result
      result(@environment.windows_interop? ? :ok : :fail, 'Windows executable interoperability is enabled',
             'Windows executable interoperability is disabled')
    end

    def load_config(results)
      config = @loader.load
      results << Result.new(*container_result(config))
      config
    rescue ConfigError => e
      results << Result.new(:fail, e.message)
      nil
    end

    # dependencies.<container> having an empty image is already a load-time
    # ConfigError (validate_dependency!), so the only way to reach here with a
    # broken primary container is `container:` naming an entry that isn't defined
    # — or, under compose-native, compose.service naming a service compose.yml
    # doesn't actually define.
    def container_result(config)
      return [:ok, 'Loaded wip.yml'] if config.compose? || config.primary
      if config.compose_native?
        return [:fail,
                "compose.service '#{config.container}' has no matching service in compose.yml"]
      end

      [:fail, "No dependencies.#{config.container} entry"]
    end

    def check_config(config, results)
      return unless config

      check_wslc(config, results)
      check_compose(config, results) if config.compose?
      check_compose_native(config, results) if config.compose_native?
      check_sync(config, results) if config.sync?
    end

    def check_wslc(config, results)
      command = resolve(config, results)
      check_version(command, results) if command
    end

    def resolve(config, results)
      command = @resolver.resolve(config.wslc_command)
      results << Result.new(:ok, "Found #{command}")
      command
    rescue CommandNotFoundError => e
      results << Result.new(:fail, e.message)
      nil
    end

    def check_version(command, results, label: 'WSLC')
      _output, status = Open3.capture2e(command, 'version')
      results << Result.new(status.success? ? :ok : :fail,
                            status.success? ? "#{label} is available" : "#{label} version failed")
    rescue Errno::ENOENT
      results << Result.new(:fail, "#{label} version failed")
    end

    def check_compose(config, results)
      command = resolve_compose(config, results)
      check_version(command, results, label: 'compose command') if command
      check_compose_file(config, results)
    end

    def resolve_compose(config, results)
      command = @compose_resolver.resolve(config.compose_command)
      results << Result.new(:ok, "Found #{command}")
      command
    rescue CommandNotFoundError => e
      results << Result.new(:fail, e.message)
      nil
    end

    # Returns whether the file was found, so check_compose_native can skip
    # attempting to parse a file that's already been reported missing.
    def check_compose_file(config, results)
      path = ComposeBridge.file_path(config)
      found = path.file?
      results << result(found ? :ok : :fail, "Found compose file #{path}", "Compose file not found: #{path}")
      found
    rescue ConfigError => e
      results << Result.new(:fail, e.message)
      false
    end

    # mode: compose-native has no external binary to check (check_wslc, above, already
    # covers the one binary it drives) — just that compose.yml exists and parses.
    def check_compose_native(config, results)
      return unless check_compose_file(config, results)

      ComposeFile.load(ComposeBridge.file_path(config))
      results << Result.new(:ok, 'Parsed compose file')
    rescue ConfigError => e
      results << Result.new(:fail, e.message)
    end

    def check_sync(config, results)
      sync = config.sync
      results << result(File.directory?(sync.source) ? :ok : :fail,
                        "Sync source #{sync.source} mirrors into volume #{sync.volume} at #{sync.target}",
                        "Sync source not found: #{sync.source}")
      return unless sync.exec? && (sync.image || sync.build)

      results << Result.new(:warn, 'sync.image/sync.build only cover `wip up`’s one-time pre-boot mirror ' \
                                   '(the primary container isn’t running yet, so that step always uses a ' \
                                   'throwaway container) — sync.mode: exec’s `wip sync`/`wip sync --watch` run ' \
                                   'rsync inside the primary container instead, so its image needs rsync too')
    end

    def result(condition, ok_message, fail_message)
      Result.new(condition,
                 condition == :ok ? ok_message : fail_message)
    end

    def command_available?(name) = ENV.fetch('PATH', '').split(File::PATH_SEPARATOR).any? { |dir| File.executable?(File.join(dir, name)) }
  end
end

File: src/Wip.Core/Diagnostics/Doctor.cs
using Wip.Compose;
using Wip.Configuration;
using Wip.Execution;
using Wip.Platform;

namespace Wip.Diagnostics;

/// <summary>Runs environment diagnostics and reports pass/warn/fail results.</summary>
public sealed class Doctor
{
    public enum Level
    {
        Ok,
        Warn,
        Fail,
    }

    public readonly record struct Result(Level Level, string Message);

    private readonly ConfigLoader loader;
    private readonly CommandResolver resolver;
    private readonly CommandResolver composeResolver;
    private readonly IEnvironment environment;

    public Doctor(
        ConfigLoader loader,
        IEnvironment environment,
        CommandResolver? resolver = null,
        CommandResolver? composeResolver = null)
    {
        this.loader = loader;
        this.environment = environment;
        this.resolver = resolver ?? new CommandResolver();
        this.composeResolver = composeResolver ??
                               new CommandResolver([], "compose command", ComposeBridge.InstallHint);
    }

    public IReadOnlyList<Result> Call()
    {
        var results = new List<Result>
        {
            new(environment.IsWsl2 ? Level.Ok : Level.Fail,
                environment.IsWsl2 ? "WSL2 is available" : "WSL2 is not available"),
            new(Level.Ok, $"Architecture: {environment.Architecture}"),
        };

        var config = LoadConfig(results);
        CheckConfig(config, results);

        results.Add(new Result(
            CommandAvailable("git") ? Level.Ok : Level.Warn,
            CommandAvailable("git") ? "Git is available" : "Git is not available to the WSLC build environment"));

        return results;
    }

    private Config? LoadConfig(List<Result> results)
    {
        try
        {
            var config = loader.Load();
            results.Add(ContainerResult(config));
            return config;
        }
        catch (ConfigException exception)
        {
            results.Add(new Result(Level.Fail, exception.Message));
            return null;
        }
    }

    /// <summary>
    /// A dependency with an empty image is already a load-time error, so the only way to get
    /// here with a broken primary container is a <c>container:</c> naming an entry that isn't
    /// defined — or, under compose-native, a <c>compose.service</c> naming a service
    /// compose.yml doesn't define.
    /// </summary>
    private static Result ContainerResult(Config config)
    {
        if (config.IsCompose || config.Primary is not null)
        {
            return new Result(Level.Ok, "Loaded wip.yml");
        }

        return config.IsComposeNative
            ? new Result(Level.Fail, $"compose.service '{config.Container}' has no matching service in compose.yml")
            : new Result(Level.Fail, $"No dependencies.{config.Container} entry");
    }

    private void CheckConfig(Config? config, List<Result> results)
    {
        if (config is null)
        {
            return;
        }

        CheckWslc(config, results);
        CheckProjectLocation(config, results);

        if (config.IsCompose)
        {
            CheckCompose(config, results);
        }

        if (config.IsComposeNative)
        {
            CheckComposeNative(config, results);
        }

        if (config.HasSync)
        {
            CheckSync(config, results);
        }
    }

    private void CheckWslc(Config config, List<Result> results)
    {
        if (Resolve(resolver, config.WslcCommand, results) is { } command)
        {
            CheckVersion(command, results);
        }
    }

    private static string? Resolve(CommandResolver resolver, string configured, List<Result> results)
    {
        try
        {
            var command = resolver.Resolve(configured);
            results.Add(new Result(Level.Ok, $"Found {command}"));
            return command;
        }
        catch (CommandNotFoundException exception)
        {
            results.Add(new Result(Level.Fail, exception.Message));
            return null;
        }
    }

    private static void CheckVersion(string command, List<Result> results, string label = "WSLC")
    {
        results.Add(ProcessProbe.Succeeds(command, ["version"])
            ? new Result(Level.Ok, $"{label} is available")
            : new Result(Level.Fail, $"{label} version failed"));
    }

    private void CheckCompose(Config config, List<Result> results)
    {
        if (Resolve(composeResolver, config.ComposeCommand ?? "auto", results) is { } command)
        {
            CheckVersion(command, results, "compose command");
        }

        CheckComposeFile(config, results);
    }

    /// <summary>
    /// Returns whether the file was found, so the compose-native check can skip parsing a
    /// file already reported missing.
    /// </summary>
    private static bool CheckComposeFile(Config config, List<Result> results)
    {
        try
        {
            var path = ComposeBridge.FilePath(config.Path ?? ".", config.ComposeFilePath);
            var found = File.Exists(path);
            results.Add(new Result(
                found ? Level.Ok : Level.Fail,
                found ? $"Found compose file {path}" : $"Compose file not found: {path}"));
            return found;
        }
        catch (ConfigException exception)
        {
            results.Add(new Result(Level.Fail, exception.Message));
            return false;
        }
    }

    /// <summary>
    /// compose-native has no external binary to check — the wslc check above already covers
    /// the one binary it drives — just that compose.yml exists and parses.
    /// </summary>
    private static void CheckComposeNative(Config config, List<Result> results)
    {
        if (!CheckComposeFile(config, results))
        {
            return;
        }

        try
        {
            ComposeFile.Load(ComposeBridge.FilePath(config.Path ?? ".", config.ComposeFilePath));
            results.Add(new Result(Level.Ok, "Parsed compose file"));
        }
        catch (ConfigException exception)
        {
            results.Add(new Result(Level.Fail, exception.Message));
        }
    }

    /// <summary>
    /// A project on the WSL filesystem reaches wip as a UNC path, and wslc resolves a
    /// bind-mount source as a Windows path — mounting an empty directory rather than
    /// failing when it does not exist. <c>sync.source</c> says so itself by refusing to
    /// resolve, but <c>volumes:</c> entries are passed to wslc verbatim, so nothing else
    /// would report the one thing that explains an empty container.
    /// </summary>
    private static void CheckProjectLocation(Config config, List<Result> results)
    {
        if (config.Path is null)
        {
            return;
        }

        var directory = Path.GetDirectoryName(Path.GetFullPath(config.Path));
        if (directory is null || !WslPath.IsWslPath(directory))
        {
            return;
        }

        results.Add(new Result(Level.Fail,
            $"Project is on the WSL filesystem ({directory}): wslc bind mounts " +
            "(volumes:, sync.source) cannot reach it, and mount empty rather than " +
            "failing. Move the project onto the Windows filesystem, e.g. C:\\src\\myproject"));
    }

    private static void CheckSync(Config config, List<Result> results)
    {
        var sync = config.Sync!;

        // Resolving the source is what refuses a WSL-side path, and doctor is exactly where
        // that has to read as a result rather than as wip falling over.
        string source;
        try
        {
            source = sync.Source;
        }
        catch (ConfigException exception)
        {
            results.Add(new Result(Level.Fail, exception.Message));
            return;
        }

        var found = Directory.Exists(source);
        results.Add(new Result(
            found ? Level.Ok : Level.Fail,
            found
                ? $"Sync source {source} mirrors into volume {sync.Volume} at {sync.Target}"
                : $"Sync source not found: {source}"));

        if (!sync.IsExec || (sync.Image is null && sync.Build is null))
        {
            return;
        }

        results.Add(new Result(Level.Warn,
            "sync.image/sync.build only cover `wip up`’s one-time pre-boot mirror " +
            "(the primary container isn’t running yet, so that step always uses a " +
            "throwaway container) — sync.mode: exec’s `wip sync`/`wip sync --watch` run " +
            "rsync inside the primary container instead, so its image needs rsync too"));
    }

    private static bool CommandAvailable(string name)
    {
        var path = System.Environment.GetEnvironmentVariable("PATH") ?? string.Empty;
        var extensions = OperatingSystem.IsWindows()
            ? (System.Environment.GetEnvironmentVariable("PATHEXT") ?? ".COM;.EXE;.BAT;.CMD")
                .Split(';', StringSplitOptions.RemoveEmptyEntries)
            : [string.Empty];

        return path.Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries)
            .Any(directory => extensions.Any(extension => File.Exists(Path.Combine(directory, name + extension))));
    }
}
